feat: account change loading state

lujakob opened the pull request:
### Describe the changes you have made in this PR

When loading an account on open popup or manual account switch data elements show a loading indicator now.

### Link this PR to an issue [optional]

Fixes #2407

### Type of change

(Remove other not matching type)

- `feat`: New feature (non-breaking change which adds functionality)

### Screenshots of the changes [optional]

_Add screenshots to make your changes easier to understand. You can also add a video here._

### How has this been tested?

Manually

### Checklist

- [ ] My code follows the style guidelines of this project and performed a self-review of my own code
- [ ] New and existing tests pass locally with my changes
- [ ] I checked if I need to make corresponding changes to the documentation (and made those changes if needed)

lujakob commented:
@reneaaron this is a working draft. Would be cool if you could have a look at this. I implemented it already three weeks ago and created the PR now that the wallet page is in master. You can test it by setting a custom Network panel "extra slow" throttling option. Also you can see in the code comments that I out-commented the cache, for the throttled request to work.

I basically change the AccountContext "loading" to "statusLoading" (for the loading of the initial status it's used) and added a distinct "accountLoading".

I added a Skeleton to the AccountMenu component, the BalanceBox and the TransactionsTable. I'm not sure about the latter, maybe a loading spinner would be nicer here.

Just switch back and forth between accounts and you should see the loading states.

Let me know what you think and if you have any general implementation improvements. Thanks.

bitcoinuser commented:
Maybe related? https://github.com/getAlby/lightning-browser-extension/issues/2463

lujakob commented:
> @reneaaron this is a working draft. Would be cool if you could have a look at this. I implemented it already three weeks ago and created the PR now that the wallet page is in master. You can test it by setting a custom Network panel "extra slow" throttling option. Also you can see in the code comments that I out-commented the cache, for the throttled request to work.
> 
> I basically change the AccountContext "loading" to "statusLoading" (for the loading of the initial status it's used) and added a distinct "accountLoading".
> 
> I added a Skeleton to the AccountMenu component, the BalanceBox and the TransactionsTable. I'm not sure about the latter, maybe a loading spinner would be nicer here.
> 
> Just switch back and forth between accounts and you should see the loading states.
> 
> Let me know what you think and if you have any general implementation improvements. Thanks.

Hi @reneaaron you merge master. Did you find time to check this PR?

reneaaron commented:
Sorry for the long wait. I indeed had a first look at it last week and will add some more technical feedback in a minute. Here are the UI-related things I stumbled upon: 

### UI related feedback

- [ ]  The boxes seem to be missing a padding
![image](https://github.com/getAlby/lightning-browser-extension/assets/100827540/6c96e43e-8010-45fd-8770-345b324e8254)

 - [ ] The skeleton loaders do not exactly fit the content they represent (so there is a "bump"  whenever the actual data is available)

 - [ ] The skeleton loader on the wallet page are indeed a bit much, I'd also go for a loading spinner here

 - [ ] The loading state seems to be missing in the account menu (while the balance is loaded, the old balance is still being displayed)
![image](https://github.com/getAlby/lightning-browser-extension/assets/100827540/308957de-5b76-4ad3-976d-bf13047cd91c)

reneaaron commented:
The incoming transaction list seems to keep loading forever now: :thinking: 

![image](https://github.com/getAlby/lightning-browser-extension/assets/100827540/30b8a6bb-ff01-4c74-bde4-675f5399a333)

lujakob commented:
@reneaaron Thanks for reviewing. I fixed all your issues. However I have one doubt.

>  The skeleton loaders do not exactly fit the content they represent (so there is a "bump" whenever the actual data is available)

Balances and accounts names have flexible widths, hence the Skeleton can never fit the exact width, as it is not yet known during the loading state. So I cannot improve this. Unless I misunderstood it, then let me know.

reneaaron commented:
> Balances and accounts names have flexible widths, hence the Skeleton can never fit the exact width, as it is not yet known during the loading state. So I cannot improve this. Unless I misunderstood it, then let me know.

It's not about the width, but the height - I just tested again and that seems to be fine now :+1: 

Will continue with the review later this week.

lujakob commented on src/app/context/AccountContext.tsx:
```diff
@@ -167,9 +182,10 @@ export function AccountProvider({ children }: { children: React.ReactNode }) {
     fetchAccountInfo,
-    loading,
-    balanceLoading,
+    statusLoading,
```
`statusLoading` is the original `loading`, it is set in https://github.com/lujakob/lightning-browser-extension/blob/fde9d695c3535252de25abe228eee6e308abf51b/src/app/context/AccountContext.tsx#L156-L157 (only on mount) and used in https://github.com/lujakob/lightning-browser-extension/blob/fde9d695c3535252de25abe228eee6e308abf51b/src/app/router/RequireAuth.tsx#L8-L9

`balanceLoading` is the new one that is used on account change.

I'm not 100% sure anymore, but I think RequireAuth line 8 returns null and nothing is displayed if I used the original, now renamed `statusLoading` for account change. Does it make sense to you?

rolznz replied:
I looked into this when updating the SWR PR - each time you re-open the popup (or any other root page) it will query the background script to get the status which includes whether or not the extension has been unlocked or not (when you close the popup all state there is destroyed). Since the fiat balance requires an API call it needs a separate loading state, but I think that would also be nice to move to SWR.

lujakob commented:
@reneaaron before merging into master (when testing is finished) we have to enable caching and uncomment https://github.com/lujakob/lightning-browser-extension/blob/fcb57bdcf8d7b0aa421d047fb38ab83af1666296/src/common/lib/api.ts#L73-L77

reneaaron reviewed (approved):
tACK

reneaaron commented:
Noticed a race condition (unrelated to this PR) that will be tackled with #2017 

https://github.com/lujakob/lightning-browser-extension/blob/fix/account-change-loading-state/src/app/context/AccountContext.tsx#L106-L107

1. Switch account to A
2. SWR resolves immediately and does the revalidate in the background
3. Current account selected is A
4. Switch account to B
5. The SWR callback is called and sets the account to A
